### PR TITLE
Respect Content-Type from internal headers instead of caching it in body

### PR DESCRIPTION
**Pull request by CYBAI**
#### cd292c4f555814e858cf649225cf2d26f036e58c
<pre>
Respect Content-Type from internal headers instead of caching it in body
<a href="https://bugs.webkit.org/show_bug.cgi?id=220535">https://bugs.webkit.org/show_bug.cgi?id=220535</a>

Reviewed by Youenn Fablet.

Previously, Content-Type is cached as `m_contentType` in
FetchBodyOwner. However, as the spec says, Content-Type should be
derived from the one in Headers. Thus, we&apos;d like to remove the cached
Content-Type and always retrieve it from internal headers instead.

* LayoutTests/http/wpt/fetch/fetch-as-blob-expected.txt: Removed.
* LayoutTests/http/wpt/fetch/fetch-as-blob-worker-expected.txt: Removed.
* LayoutTests/http/wpt/fetch/fetch-as-blob-worker.html: Removed.
* LayoutTests/http/wpt/fetch/fetch-as-blob.html: Removed.
* LayoutTests/http/wpt/fetch/fetch-as-blob.js: Removed.
* LayoutTests/imported/w3c/resources/import-expectations.json:
* LayoutTests/imported/w3c/web-platform-tests/fetch/api/body/cloned-any.js: Added.
(promise_test.async t):
* LayoutTests/imported/w3c/web-platform-tests/fetch/api/body/mime-type.any-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/fetch/api/body/mime-type.any.js:
(forEach.bodyContainerCreator.promise_test.async t):
(forEach.bodyContainerCreator.assert_equals):
(new.Response.new.Blob.string_appeared_here.forEach.bodyContainerCreator.promise_test.async t):
(new.Response.new.Blob.string_appeared_here.forEach.bodyContainerCreator.assert_equals):
(new.Response.new.Blob):
* LayoutTests/imported/w3c/web-platform-tests/fetch/api/body/mime-type.any.worker-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/fetch/api/body/w3c-import.log:
* LayoutTests/imported/w3c/web-platform-tests/service-workers/cache-storage/cache-match.https.any-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/service-workers/cache-storage/cache-match.https.any.serviceworker-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/service-workers/cache-storage/cache-match.https.any.sharedworker-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/service-workers/cache-storage/cache-match.https.any.worker-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/service-workers/cache-storage/serviceworker/cache-match.https-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/service-workers/cache-storage/window/cache-match.https-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/service-workers/cache-storage/worker/cache-match.https-expected.txt:
* Source/WebCore/Modules/fetch/FetchBody.cpp:
(WebCore::FetchBody::blob):
(WebCore::FetchBody::consumeOnceLoadingFinished):
* Source/WebCore/Modules/fetch/FetchBody.h:
* Source/WebCore/Modules/fetch/FetchBodyConsumer.cpp:
(WebCore::resolveWithTypeAndData):
(WebCore::FetchBodyConsumer::resolve):
(WebCore::FetchBodyConsumer::takeAsBlob):
(WebCore::FetchBodyConsumer::clone):
* Source/WebCore/Modules/fetch/FetchBodyConsumer.h:
(WebCore::FetchBodyConsumer::setContentType): Deleted.
* Source/WebCore/Modules/fetch/FetchBodyOwner.cpp:
(WebCore::FetchBodyOwner::blob):
(WebCore::FetchBodyOwner::cloneBody):
(WebCore::FetchBodyOwner::extractBody):
(WebCore::FetchBodyOwner::consumeOnceLoadingFinished):
(WebCore::FetchBodyOwner::formData):
(WebCore::FetchBodyOwner::updateContentType): Deleted.
* Source/WebCore/Modules/fetch/FetchBodyOwner.h:
(WebCore::FetchBodyOwner::contentType const):
* Source/WebCore/Modules/fetch/FetchRequest.cpp:
(WebCore::m_signal):
(WebCore::FetchRequest::initializeWith):
* Source/WebCore/Modules/fetch/FetchResponse.cpp:
(WebCore::FetchResponse::create):
(WebCore::FetchResponse::clone):
(WebCore::FetchResponse::setReceivedInternalResponse):

Canonical link: <a href="https://commits.webkit.org/264054@main">https://commits.webkit.org/264054@main</a>
</pre>

https://github.com/WebKit/WebKit/commit/58c80f65289a23ad03d48a8ba6b7d6d6665e2202

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/6541 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/6753 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/6933 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/8122 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/6825 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/6538 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/7220 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/6702 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/9722 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/6656 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/6807 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/5940 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/8210 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/4268 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/5910 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/13748 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/6237 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/5990 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/8386 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/6474 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/5296 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/5875 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/1544 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/10039 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/6247 "Built successfully") | | |

### CODE REVIEW

File: Source/WebCore/Modules/fetch/FetchBodyOwner.cpp
Comment (CYBAI) [inline Source/WebCore/Modules/fetch/FetchBodyOwner.cpp]: @youennf @annevk Question here 🙋 

Previously, we're passing `m_contentType` as a reference to `FetchBody::extract`'s second argument and mutate it directly when necessary. However, as I change the `contentType` function to _not_ return the value as a reference, it wouldn't update to the m_headers and I need to do the `m_headers->fastSet` here. Would you think it's better to return a pair of `(Body, NewContentType)` from the extract function instead? I wonder it would be helpful to remind the next person who will use the function that the content type would potentially be updated.

File: Source/WebCore/Modules/fetch/FetchBodyOwner.cpp
Comment (annevk) [inline Source/WebCore/Modules/fetch/FetchBodyOwner.cpp]: Nit: I suspect this works:
```suggestion
    auto currentContentType = contentType();
```

File: Source/WebCore/Modules/fetch/FetchBodyOwner.cpp
Comment (annevk) [reply Source/WebCore/Modules/fetch/FetchBodyOwner.cpp]: I'm not sure I understand the question. Doesn't `FetchBody::extract` essentially return a pair? One of them via an output parameter? This looks okay to me.

**Review by annevk — COMMENTED** [review]
Nice, thanks for working on this!

File: Source/WebCore/Modules/fetch/FetchBodyConsumer.cpp
Comment (youennf) [inline Source/WebCore/Modules/fetch/FetchBodyConsumer.cpp]: &contentType.
Also we can probably removing capturing context.
```
promise->resolveCallbackValueWithNewlyCreated<IDLInterface<Blob>>([this, &contentType](auto& context) {
```

File: Source/WebCore/Modules/fetch/FetchBodyOwner.cpp
Comment (youennf) [inline Source/WebCore/Modules/fetch/FetchBodyOwner.cpp:132]: We can probably do some additional refactoring here: remove the context variable and get it from the function.
```
promise->resolveCallbackValueWithNewlyCreated<IDLInterface<Blob>>([this](auto& context) {
```

File: Source/WebCore/Modules/fetch/FetchBodyOwner.cpp
Comment (youennf) [inline Source/WebCore/Modules/fetch/FetchBodyOwner.cpp]: We probably do not need this change.

File: Source/WebCore/Modules/fetch/FetchBodyOwner.cpp
Comment (youennf) [reply Source/WebCore/Modules/fetch/FetchBodyOwner.cpp]: Yeah, we could probably a later refactoring to have FetchBody::extract return all out parameters as return parameters. Probably not in this patch though.

Comment (CYBAI) [reply Source/WebCore/Modules/fetch/FetchBodyOwner.cpp]: > I'm not sure I understand the question. Doesn't `FetchBody::extract` essentially return a pair? One of them via an output parameter? This looks okay to me.

oh, you are right 🙏  I didn't realize we can see it as an output parameter.

> Yeah, we could probably a later refactoring to have FetchBody::extract return all out parameters as return parameters. Probably not in this patch though.

yes, please let me know if you think we should refactor it and I will file an issue for it after this PR is merged 🙇

File: Source/WebCore/Modules/fetch/FetchBodyOwner.cpp
Comment (CYBAI) [reply Source/WebCore/Modules/fetch/FetchBodyOwner.cpp]: ah. good catch! I will change it back 🙏

File: Source/WebCore/Modules/fetch/FetchBodyOwner.cpp
Comment (CYBAI) [reply Source/WebCore/Modules/fetch/FetchBodyOwner.cpp:132]: thanks for the tips! I will fix the capturing context in upcoming commits.

**Comment by CYBAI** [comment]
As there are some unexpected WPT failures, I will look into them and update the PR lately.

File: Source/WebCore/Modules/fetch/FetchBodyOwner.cpp
Comment (CYBAI) [inline Source/WebCore/Modules/fetch/FetchBodyOwner.cpp]: I end up realizing that we should only update the content type to headers when it was empty before running `FetchBody::extract`. If any existing content-type already exists in headers, we should keep it as is.

**Comment by CYBAI** [comment]
As more tests pass with this patch, I've marked them as PASS in the latest (9085c92) commit. For some other failures, let me try again to see the test results; some of them are not reproducible on my end :/

However, there's one failure that I need to confirm with @annevk and @youennf 

in this patch, webkit fail to pass `blob.type` assertion in this test 👇 

https://github.com/WebKit/WebKit/blob/512db9a6900b400f9ce93e90e6e178fa74fef03e/LayoutTests/http/wpt/fetch/fetch-as-blob.js#L62-L73

before this patch, webkit will use the "type" from Blob class so the test was PASS; however, from this patch, webkit will start to use the Content-Type from headers so it starts to fail instead. After reading https://fetch.spec.whatwg.org/#dom-body-blob, the spec says `blob()` will

> return a [Blob](https://w3c.github.io/FileAPI/#dfn-Blob) whose contents are bytes and whose [type](https://w3c.github.io/FileAPI/#dfn-type) attribute is [this](https://webidl.spec.whatwg.org/#this)’s [MIME type](https://fetch.spec.whatwg.org/#concept-body-mime-type).

it seems spec doesn't mention how to handle the case when `type` in Blob exists and only says "type is the request/response 's MIME type" so it sounds correct to me if we should respect the request/response's Content-Type. Also, I tried to check that Chrome will return `type` of the Blob and Firefox will return `Content-Type` of the request/response.

should I file a spec issue to discuss this? or, should I fix the test or the patch in this PR? 🙏

File: Source/WebCore/Modules/fetch/FetchBodyOwner.cpp
Comment (annevk) [reply Source/WebCore/Modules/fetch/FetchBodyOwner.cpp]: Note that empty and null are not the same thing. We probably want to call this contentTypeIsMissing or some such as that's different from it being the empty string.

**Comment by annevk** [comment]
You should update the test and maybe also add WPT coverage for that case if there is none currently. The spec indeed requires that in that case the `Content-Type` header wins due to "this's MIME type" being an algorithm that computes it based on the headers. (I could see us changing the specification to make that even more clear though, so if you think it should be clarified more feel free to file an issue.)

File: Source/WebCore/Modules/fetch/FetchBodyOwner.cpp
Comment (CYBAI) [reply Source/WebCore/Modules/fetch/FetchBodyOwner.cpp]: ah, good point! Thanks for catching this. Let me rename the variable.

**Comment by CYBAI** [comment]
> You should update the test and maybe also add WPT coverage for that case if there is none currently.

Thanks for the confirmation! I will try to move the test from `http/wpt/fetch/fetch-as-blob.js` to another related WPT test (or create a new one if necessary).

> The spec indeed requires that in that case the `Content-Type` header wins due to "this's MIME type" being an algorithm that computes it based on the headers. (I could see us changing the specification to make that even more clear though, so if you think it should be clarified more feel free to file an issue.)

yes! I would want to file an issue for it. I will find some time to do it 🙏 thanks for the clarification.

File: LayoutTests/http/wpt/fetch/fetch-as-blob.js
Comment (CYBAI) [inline LayoutTests/http/wpt/fetch/fetch-as-blob.js]: I just updated the test here to confirm there's no other issues. But, I've sent a PR to add this test to upstream wpt so I will import that upstream tests and remove this one when the upstream PR is merged.

**Comment by CYBAI** [comment]
uh, the failures are caused by accidentally removing last newline during rebase 😅

File: LayoutTests/http/wpt/fetch/fetch-as-blob.js
Comment (annevk) [reply LayoutTests/http/wpt/fetch/fetch-as-blob.js]: I've reviewed and merged https://github.com/web-platform-tests/wpt/pull/39489.

The WebKit changes look reasonable to me, but @youennf will need to do a final review.

Comment (CYBAI) [reply LayoutTests/http/wpt/fetch/fetch-as-blob.js]: Thank you so much for your help to merge the WPT PR 🙏 
I will import them and update this PR later 🙇

File: LayoutTests/http/wpt/fetch/fetch-as-blob.js
Comment (CYBAI) [inline LayoutTests/http/wpt/fetch/fetch-as-blob.js:5]: These tests in this file are already merged into the `fetch/api/body/mime-type.any.js` WPT tests so this one is removed in this PR.

**Comment by CYBAI** [comment]
@youennf the test is imported from WPT now so this is ready for review 🙇

File: Source/WebCore/Modules/fetch/FetchBodyConsumer.cpp
Comment (youennf) [inline Source/WebCore/Modules/fetch/FetchBodyConsumer.cpp]: It is only used once, let's call packageFormData with `promise->scriptExecutionContext()` directly

File: Source/WebCore/Modules/fetch/FetchBodyOwner.cpp
Comment (youennf) [inline Source/WebCore/Modules/fetch/FetchBodyOwner.cpp]: Why is this needed?
We are only supposed to clone the body here, callers of cloneBody should clone the headers themselves?
This might be to handle FetchResponse::clone, in which case we might actually want to change FetchResponse::clone.

File: Source/WebCore/Modules/fetch/FetchBodyOwner.cpp
Comment (youennf) [inline Source/WebCore/Modules/fetch/FetchBodyOwner.cpp]: wasEmptyContentType is not a great name, the isNull check means there was no content-type header, it does not mean there was a content-type header with an empty value.

**Review by youennf — COMMENTED** [review]
This looks almost ready to me and is a nice simplification.
Some questions/suggestions below.

File: Source/WebCore/Modules/fetch/FetchBodyOwner.cpp
Comment (CYBAI) [reply Source/WebCore/Modules/fetch/FetchBodyOwner.cpp]: no, thanks for catching this! I will move this logic to `FetchResponse::clone` and leave this function only take care of the clone of body 🙏 (we don't need to clone headers in `FetchRequest::clone` because it already pass the headers to `FetchRequest::create`)

File: Source/WebCore/Modules/fetch/FetchBodyOwner.cpp
Comment (CYBAI) [reply Source/WebCore/Modules/fetch/FetchBodyOwner.cpp]: you're right. please let me rename it to `contentTypeNotExist` 🙏

File: Source/WebCore/Modules/fetch/FetchBodyOwner.cpp
Comment (youennf) [inline Source/WebCore/Modules/fetch/FetchBodyOwner.cpp]:
```suggestion
    bool isContentTypeSet = !currentContentType.isNull();
```

**Review by youennf — APPROVED** [review]

File: Source/WebCore/Modules/fetch/FetchBodyOwner.cpp
Comment (youennf) [reply Source/WebCore/Modules/fetch/FetchBodyOwner.cpp]: And then need to change `contentTypeNotExist ` by `!isContentTypeSet` below

Comment (CYBAI) [reply Source/WebCore/Modules/fetch/FetchBodyOwner.cpp]: thanks for the suggestion 🙏

**Comment by CYBAI** [comment]
The WPT failure is about WebGL and I think it's not related to this PR though, I just rebased to latest main (8c497617ec2a) and let the CI to run again.

**Comment by Ahmad-S792** [comment]
> The WPT failure is about WebGL and I think it's not related to this PR though, I just rebased to latest main ([8c49761](https://github.com/WebKit/WebKit/commit/8c497617ec2a904ffb93e425f5b6ca882e1fe56a)) and let the CI to run again.

Do you have commit rights, if not, should I commit it for you?

**Comment by CYBAI** [comment]
> Do you have commit rights, if not, should I commit it for you?

No, please help me commit this patch, thank you! 🙏

**Comment by CYBAI** [comment]
Thanks for Anne and Youenn helping me a lot on this fix! and @Ahmad-S792 thank you again for helping me merge it :)

**Pull request merged**
